x/xcvm/keeper/transfer_intent.go

```diff
@@ -0,0 +1,448 @@
+package keeper
+
+import (
+	"crypto/ecdsa"
+	"encoding/binary"
+	"fmt"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/store"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/gogoproto/proto"
+	ibccore "github.com/cosmos/ibc-go/v7/modules/core/exported"
+	wasmtypes "github.com/cosmos/ibc-go/v7/modules/light-clients/08-wasm/types"
+	"github.com/ethereum/go-ethereum/common"
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/notional-labs/composable/v6/x/xcvm/types"
+	prysmtypes "github.com/prysmaticlabs/prysm/v4/proto/eth/v1"
+	"math/big"
+	"strconv"
+)
+
+// SendEthTransferIntent creates and stores a new transfer intent and emits an event to alert potential solvers
+func (k Keeper) SendEthTransferIntent(ctx sdk.Context, msg *types.MsgSendTransferIntent) error {
+	clientId := msg.ClientId
+
+	if err := k.validateClientState(ctx, clientId); err != nil {
+		return fmt.Errorf("validate client state: %v", err)
+	}
+
+	intentId := k.getNextIntentId(ctx)
+	transferIntent := types.TransferIntent{
+		ClientId:           clientId,
+		SourceAddress:      msg.FromAddress,
+		DestinationAddress: msg.DestinationAddress,
+		TimeoutHeight:      msg.TimeoutHeight,
+		TransferTokens:     msg.TransferTokens,
+		Bounty:             msg.Bounty,
+	}
+	k.AddTransferIntent(ctx, transferIntent, intentId)
+
+	userAddress, err := sdk.AccAddressFromBech32(msg.FromAddress)
+	if err != nil {
+		return fmt.Errorf("acc address conversion: %v", err)
+	}
+	coins := sdk.NewCoins(msg.Bounty)
+
+	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, userAddress, types.ModuleName, coins); err != nil {
+		return fmt.Errorf("send coins from account to module: %v", err)
+	}
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventAddTransferIntent,
+		sdk.NewAttribute(types.AttributeKeyIntentId, strconv.FormatUint(intentId, 10)),
+		sdk.NewAttribute(types.AttributeKeyClientId, transferIntent.ClientId),
+		sdk.NewAttribute(types.AttributeKeySourceAddress, transferIntent.SourceAddress),
+		sdk.NewAttribute(types.AttributeKeyDestinationAddress, transferIntent.DestinationAddress),
+		sdk.NewAttribute(types.AttributeKeyTimeout, strconv.FormatInt(transferIntent.TimeoutHeight, 10)),
+		sdk.NewAttribute(types.AttributeKeyAmount, transferIntent.TransferTokens.String()),
+		sdk.NewAttribute(types.AttributeKeyBounty, transferIntent.Bounty.String()),
+	))
+
+	k.SetNextIntentId(ctx, intentId+1)
+
+	return nil
+}
+
+// getNextIntentId retrieves the next intent id from the store
+func (k Keeper) getNextIntentId(ctx sdk.Context) uint64 {
+	kvStore := ctx.KVStore(k.storeKey)
+
+	intentIdBz := kvStore.Get(types.TransferIntentIdKey)
+	var intentId uint64
+	if intentIdBz == nil {
+		intentId = 0
+	} else {
+		intentId = binary.BigEndian.Uint64(intentIdBz)
+	}
```

Comment:
nit: unnecessary branch, default value of `uint64` is zero
```suggestion
	var intentId uint64
	if intentIdBz := kvStore.Get(types.TransferIntentIdKey); intentIdBz != nil {
		intentId = binary.BigEndian.Uint64(intentIdBz)
	}
```